### PR TITLE
Backport: WP_Theme_JSON_Resolver changes

### PR DESCRIPTION
**Pull request by Mamaduka**
Backport changes and fixes merged into a Gutenberg plugin after WP 6.1.

Related Gutenberg PRs:
* https://github.com/WordPress/gutenberg/pull/46554
* https://github.com/WordPress/gutenberg/pull/46043
* https://github.com/WordPress/gutenberg/pull/45969

I'm going to backport https://github.com/WordPress/gutenberg/pull/46112 separately.

Trac ticket: https://core.trac.wordpress.org/ticket/57545

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 1]
```diff
@@ -970,7 +970,7 @@ public function test_get_style_variations_returns_all_variations() {
 		wp_recursive_ksort( $actual_settings );
 		wp_recursive_ksort( $expected_settings );
 
-		$this->assertSame(
+		$this->assertSameSets(
```
Comment (Mamaduka) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]: Tests are failing on CI with `assertSame`, but they're passing for me locally.

```
npm run test:php -- --filter Tests_Theme_wpThemeJsonResolver
```

It's different tests that fails 😅

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 2]
```diff
@@ -544,8 +552,8 @@ public static function get_user_data() {
 	 *              added the `$origin` parameter.
 	 * @since 6.1.0 Added block data and generation of spacingSizes array.
```
Comment (audrasjb) [inline src/wp-includes/class-wp-theme-json-resolver.php:553]: Maybe add a `since` mention here?
`@since 6.2.0 Changed ' $origin' parameter values to 'default', 'blocks', 'theme' or 'custom'.`

**Review by audrasjb — COMMENTED** [review]
small docblock change proposal

File: src/wp-includes/class-wp-theme-json-resolver.php (hunk 2 above)
Comment (Mamaduka) [reply src/wp-includes/class-wp-theme-json-resolver.php:553]: Added in [0e595e6](https://github.com/WordPress/wordpress-develop/pull/3901/commits/0e595e6cb3519d6dbb0268bf8048c28dc63d4999).

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php [hunk 3]
```diff
@@ -511,6 +532,10 @@ public function test_get_theme_items() {
 				),
 			),
 		);
-		$this->assertSameSetsWithIndex( $data, $expected );
+
+		wp_recursive_ksort( $data );
+		wp_recursive_ksort( $expected );
+
+		$this->assertSameSets( $data, $expected );
```
Comment (Mamaduka) [inline tests/phpunit/tests/rest-api/rest-global-styles-controller.php:539]: Is it preferred sorting arrays before assertions, or can we use something like `assertEqualsCanonicalizing`?

Comment (hellofromtonya) [reply tests/phpunit/tests/rest-api/rest-global-styles-controller.php:539]: Here the deeply nested array of arrays are being recursively key sorted before the assertion. `assertSameSets()` repeats an index sort at the first dimension of the arrays. That's not necessary.

Instead of `assertSameSets()`, use `assertSame()`. 

I'm wondering though why this needed. Pulling down to explore.

Comment (Mamaduka) [reply tests/phpunit/tests/rest-api/rest-global-styles-controller.php:539]: Hi, @hellofromtonya

Let me know if any follow-ups are needed for PHPUnits tests on this PR.

**Comment by hellofromtonya** [comment]
Hey @Mamaduka, is this PR and Trac ticket still needed? I remember you were breaking up the GB PRs into smaller Trac tickets. Just checking before I dive into it.

**Comment by Mamaduka** [comment]
@hellofromtonya, that was for `WP_Theme_JSON` file. I decided to keep this one since I had already done the work, and most of the line changes were for PHPUnit tests.

File: src/wp-includes/class-wp-theme-json-resolver.php [hunk 4]
```diff
@@ -647,33 +665,61 @@ public static function clean_cached_data() {
 		static::$i18n_schema              = null;
 	}
 
+	/**
+	 * Returns an array of all nested json files within a given directory.
```
Comment (costdev) [inline src/wp-includes/class-wp-theme-json-resolver.php]:
```suggestion
	 * Returns an array of all nested JSON files within a given directory.
```
Nitpick: Capitalize JSON here.

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 5]
```diff
@@ -775,4 +775,204 @@ public function test_get_theme_data_does_not_parse_theme_json_if_not_present() {
...
+	 *
+	 * @param string $origin            What origin to get data from.
+	 * @param bool   $core_palette      Whether the core palette is present.
+	 * @param string $core_palette_text Message.
+	 * @param string $block_styles      Whether the block styles are present.
+	 * @param string $block_styles_text Message.
+	 * @param bool   $theme_palette      Whether the theme palette is present.
+	 * @param string $theme_palette_text Message.
+	 * @param bool   $user_palette      Whether the user palette is present.
+	 * @param string $user_palette_text Message.
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]:
```suggestion
	 * @param string $origin             What origin to get data from.
	 * @param bool   $core_palette       Whether the core palette is present.
	 * @param string $core_palette_text  Message.
	 * @param string $block_styles       Whether the block styles are present.
	 * @param string $block_styles_text  Message.
	 * @param bool   $theme_palette      Whether the theme palette is present.
	 * @param string $theme_palette_text Message.
	 * @param bool   $user_palette        Whether the user palette is present.
	 * @param string $user_palette_text   Message.
```

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 6]
```diff
@@ -775,4 +775,204 @@ public function test_get_theme_data_does_not_parse_theme_json_if_not_present() {
...
+			$styles,
+			static function( $element ) {
+				return isset( $element['name'] ) && 'my/block-with-styles' === $element['name'];
+			}
+		);
+		$this->assertSame( $block_styles, count( $styles ) === 1, $block_styles_text );
+		$this->assertSame( $theme_palette, isset( $settings['color']['palette']['theme'] ), $theme_palette_text );
+		$this->assertSame( $user_palette, isset( $settings['color']['palette']['custom'] ), $user_palette_text );
+
+		unregister_block_type( 'my/block-with-styles' );
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]: To ensure all code runs before the assertions, preventing failures from leaving anything behind accidentally, this can be changed to:
```suggestion
		$styles = array_filter(
			$styles,
			static function( $element ) {
				return isset( $element['name'] ) && 'my/block-with-styles' === $element['name'];
			}
		);
		unregister_block_type( 'my/block-with-styles' );
		$this->assertSame( $core_palette, isset( $settings['color']['palette']['default'] ), $core_palette_text );
		$this->assertSame( $block_styles, count( $styles ) === 1, $block_styles_text );
		$this->assertSame( $theme_palette, isset( $settings['color']['palette']['theme'] ), $theme_palette_text );
		$this->assertSame( $user_palette, isset( $settings['color']['palette']['custom'] ), $user_palette_text );
```
This suggestion might be a little difficult to read, but basically I just moved the `$styles = array_filter(` and `unregister_block_type()` sections above the assertions.

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php (hunk 5 above)
Comment (costdev) [reply tests/phpunit/tests/theme/wpThemeJsonResolver.php]: Alternatively, this could document an `$args` array, so that the test method just accepts a single argument instead of 9, reducing horizontal scrolling.

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 7]
```diff
@@ -775,4 +775,204 @@ public function test_get_theme_data_does_not_parse_theme_json_if_not_present() {
 		$this->assertSame( $empty_theme_json, $theme_data->get_raw_data(), 'Theme data should be empty without theme support.' );
 		$this->assertNull( $property->getValue(), 'Theme i18n schema should not have been loaded without theme support.' );
 	}
+
+	/**
+	 * Test that get_merged_data returns the data merged up to the proper origin.
+	 *
+	 * @covers WP_Theme_JSON_Resolver::get_merged_data
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php:784]:
```suggestion
	 * @ticket 57545
	 *
	 * @covers WP_Theme_JSON_Resolver::get_merged_data
```

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 8]
```diff
@@ -775,4 +775,204 @@ public function test_get_theme_data_does_not_parse_theme_json_if_not_present() {
...
+		$this->assertSame( $theme_palette, isset( $settings['color']['palette']['theme'] ), $theme_palette_text );
+		$this->assertSame( $user_palette, isset( $settings['color']['palette']['custom'] ), $user_palette_text );
+
+		unregister_block_type( 'my/block-with-styles' );
+	}
+
+	/**
+	 * Data provider for test_get_merged_data_returns_origin
+	 *
+	 * @return array
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]:
```suggestion
	 * Data provider.
	 *
	 * @return array[]
```
This is the format that we have agreed on in Core as a "for now" approach for consistency, until the community reaches consensus on a standardized format.

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 9]
```diff
@@ -775,4 +775,204 @@ public function test_get_theme_data_does_not_parse_theme_json_if_not_present() {
...
+				'user_palette_text'  => 'User palette must be present',
+			),
+		);
+	}
+
+	/**
+	 * Test that get_style_variations returns all variations, including parent theme variations if the theme is a child,
+	 * and that the child variation overwrites the parent variation of the same name.
+	 *
+	 * @covers WP_Theme_JSON_Resolver::get_style_variations
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php:917]:
```suggestion
	 * @ticket 57545
	 *
	 * @covers WP_Theme_JSON_Resolver::get_style_variations
```

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 10]
```diff
@@ -775,4 +775,204 @@ public function test_get_theme_data_does_not_parse_theme_json_if_not_present() {
 		$this->assertSame( $empty_theme_json, $theme_data->get_raw_data(), 'Theme data should be empty without theme support.' );
 		$this->assertNull( $property->getValue(), 'Theme i18n schema should not have been loaded without theme support.' );
 	}
+
+	/**
+	 * Test that get_merged_data returns the data merged up to the proper origin.
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]:
```suggestion
	 * Tests that get_merged_data returns the data merged up to the proper origin.
```

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 11]
```diff
@@ -775,4 +775,204 @@ public function test_get_theme_data_does_not_parse_theme_json_if_not_present() {
...
+				'theme_palette'      => true,
+				'theme_palette_text' => 'Theme palette must be present',
+				'user_palette'       => true,
+				'user_palette_text'  => 'User palette must be present',
+			),
+		);
+	}
+
+	/**
+	 * Test that get_style_variations returns all variations, including parent theme variations if the theme is a child,
```
Comment (costdev) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]:
```suggestion
	 * Tests that get_style_variations returns all variations, including parent theme variations if the theme is a child,
```

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the PR @Mamaduka! Just a few suggestions. 🙂

**Comment by Mamaduka** [comment]
Thanks for the suggestions, @costdev!

File: tests/phpunit/tests/theme/wpThemeJsonResolver.php [hunk 12]
```diff
@@ -775,4 +775,209 @@ public function test_get_theme_data_does_not_parse_theme_json_if_not_present() {
...
+				'slug'  => 'my-color',
+			),
+		);
+		$user_cpt['post_content']                         = wp_json_encode( $config );
+		wp_update_post( $user_cpt, true, false );
+
+		$theme_json = WP_Theme_JSON_Resolver::get_merged_data( $origin );
+		$settings   = $theme_json->get_settings();
+		$styles     = $theme_json->get_styles_block_nodes();
+		$styles = array_filter(
```
Comment (Mamaduka) [inline tests/phpunit/tests/theme/wpThemeJsonResolver.php]:
```suggestion
		$styles     = array_filter(
```

**Comment by felixarntz** [comment]
## Performance check ✅ 

I gave this some server-side performance tests for the frontend, measuring TT3 on the home page and a single post with the Gutenberg demo content, using https://gist.github.com/felixarntz/de5c697a1a16c2b892634b70216eb6c7.

### TLDR / Summary / Conclusion
Results look inconclusive. Function-specific benchmarks show a clear improvement, while overall WordPress execution time shows inconsistent results between the two tests. This suggests that the changes in this PR probably enhance performance, but overall does not have a great impact on performance. At the same time, it also means that there is no concern about merging this, so we can proceed from that perspective.

### Results

_All times below are in milliseconds._

First I benchmarked overall WP load time. Every test row below is actually based on the median of 20 runs itself. So overall this is 120 runs per scenario.

  | trunk @ [55218] | plus PR #3901 | % diff
-- | -- | -- | --
Home page (test 1) | 193.99 | 206.9 | 6.65%
Gutenberg demo (test 1) | 195.65 | 193.1 | -1.30%
Home page (test 2) | 194.31 | 198.11 | 1.96%
Gutenberg demo (test 2) | 194.98 | 191.46 | -1.81%
Home page (test 3) | 198.5 | 203.47 | 2.50%
Gutenberg demo (test 3) | 193.64 | 202.12 | 4.38%

Because the results didn't show a clear trend in either direction, I wanted to see at least more specifically whether certain `WP_Theme_JSON_Resolver` class methods showed performance increases or regressions here.

Here are the results for `WP_Theme_JSON_Resolver::get_merged_data()`, again based on 20 runs per scenario (for the home page):

  | trunk @ [55218] | plus PR #3901 | % diff
-- | -- | -- | --
get_merged_data() time | 14.63 | 13.99 | -4.37%
get_merged_data() calls | 50 | 50 | 0.00%
WP total time | 199.01 | 202.04 | 1.52%

Here are the results for `WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles()`, again based on 20 runs per scenario (for the home page):

  | trunk @ [55218] | plus PR #3901 | % diff
-- | -- | -- | --
get_user_data_from_...() time | 2.5 | 1.57 | -37.20%
get_user_data_from...() calls | 2 | 2 | 0.00%
WP total time | 198.64 | 196.6 | -1.03%

Both of the above show improvements for their respective method's performance, so that's promising. We will have to see whether this has any impact in the field, but overall it certainly looks unproblematic, and if anything, there is a positive trend in `WP_Theme_JSON_Resolver` performance. The overall WP execution time differences are too inconsistent to mean anything, likely due to high variance.

_**Note:** The `get_style_variations()` method was not covered as it is not invoked in the frontend and thus has no impact there._

**Review by felixarntz — APPROVED** [review]
@Mamaduka Thanks! Code-wise this all looks solid to me, and performance-wise as well per my above feedback.

**Review by spacedmonkey — APPROVED** [review]

**Comment by Mamaduka** [comment]
Thanks for the reviews and testing, folks 🙇

It would be great to get this committed before Beta 1.

**Comment by ntsekouras** [comment]
Committed via: https://core.trac.wordpress.org/changeset/55231
